File: lexer/AstPrinter.cs
// ============================================================
// AstPrinter.cs
// Vuelca un AST de BASICPLUS en formato indentado tipo árbol,
// útil para depurar el parser y comprobar visualmente que la
// estructura de un programa es la esperada.
// ============================================================

using System.Text;

namespace BasicPlus.Lexer;

public static class AstPrinter
{
    public static string Print(AstNode node)
    {
        var sb = new StringBuilder();
        Write(node, sb, 0);
        return sb.ToString();
    }

    private static void Write(object? node, StringBuilder sb, int depth)
    {
        string ind = new string(' ', depth * 2);
        switch (node)
        {
            case null:
                sb.AppendLine($"{ind}<null>");
                break;

            // -------- Módulo / imports --------
            case ModuleNode m:
                sb.AppendLine($"{ind}Module '{m.Name}'  @ {m.Line}:{m.Column}");
                foreach (var imp in m.Imports) Write(imp, sb, depth + 1);
                foreach (var d in m.Defs) Write(d, sb, depth + 1);
                break;
            case ImportNode imp:
                sb.AppendLine($"{ind}Import {string.Join(".", imp.Path)}");
                break;

            // -------- Tipos --------
            case SimpleTypeRef st:
                sb.AppendLine($"{ind}Type {st.Name}");
                break;
            case ArrayTypeRef at:
                sb.AppendLine($"{ind}ArrayType");
                Write(at.Element, sb, depth + 1);
                if (at.Size is not null)
                {
                    sb.AppendLine($"{ind}  size:");
                    Write(at.Size, sb, depth + 2);
                }
                break;

            // -------- Helpers --------
            case DeclName dn:
                sb.AppendLine($"{ind}DeclName {(dn.IsStatic ? dn.ClassQualifier + "." : "")}{dn.Name}");
                break;
            case Param p:
                sb.AppendLine($"{ind}Param {p.Name}");
                Write(p.Type, sb, depth + 1);
                break;
            case EnumValue ev:
                string val = ev.ExplicitValue is { } v ? $" := {v}" : "";
                sb.AppendLine($"{ind}EnumValue {ev.Name}{val}");
                break;
            case GetterDef g:
                sb.AppendLine($"{ind}Getter");
                foreach (var s in g.Body) Write(s, sb, depth + 1);
                break;
            case SetterDef s:
                sb.AppendLine($"{ind}Setter (param={s.ParamName})");
                foreach (var st in s.Body) Write(st, sb, depth + 1);
                break;

            // -------- Declaraciones --------
            case ConstDecl c:
                sb.AppendLine($"{ind}ConstDecl{Mods(c.IsPublic, false)}");
                Write(c.Name, sb, depth + 1);
                if (c.Type is not null) Write(c.Type, sb, depth + 1);
                sb.AppendLine($"{ind}  value:");
                Write(c.Value, sb, depth + 2);
                break;
            case VarDecl v:
                sb.AppendLine($"{ind}VarDecl{Mods(v.IsPublic, false)}");
                foreach (var n in v.Names) Write(n, sb, depth + 1);
                Write(v.Type, sb, depth + 1);
                if (v.Init is not null)
                {
                    sb.AppendLine($"{ind}  init:");
                    Write(v.Init, sb, depth + 2);
                }
                break;
            case FuncDef f:
                sb.AppendLine($"{ind}FuncDef{Mods(f.IsPublic, f.IsFinal)}");
                Write(f.Name, sb, depth + 1);
                foreach (var p in f.Params) Write(p, sb, depth + 1);
                if (f.ReturnType is not null)
                {
                    sb.AppendLine($"{ind}  returns:");
                    Write(f.ReturnType, sb, depth + 2);
                }
                sb.AppendLine($"{ind}  body:");
                foreach (var st in f.Body) Write(st, sb, depth + 2);
                break;
            case PropertyDef pr:
                sb.AppendLine($"{ind}PropertyDef{Mods(pr.IsPublic, pr.IsFinal)}{(pr.IsShortForm ? " [short]" : "")}");
                Write(pr.Name, sb, depth + 1);
                Write(pr.Type, sb, depth + 1);
                if (pr.Init is not null)
                {
                    sb.AppendLine($"{ind}  init:");
                    Write(pr.Init, sb, depth + 2);
                }
                if (pr.Getter is not null) Write(pr.Getter, sb, depth + 1);
                if (pr.Setter is not null) Write(pr.Setter, sb, depth + 1);
                break;
            case ClassDef cd:
                sb.AppendLine($"{ind}ClassDef '{cd.Name}'{(cd.BaseClass is null ? "" : $" extends {cd.BaseClass}")}{Mods(cd.IsPublic, false)}");
                foreach (var m in cd.Members) Write(m, sb, depth + 1);
                break;
            case EnumDef en:
                sb.AppendLine($"{ind}EnumDef '{en.Name}'{Mods(en.IsPublic, false)}");
                foreach (var ev in en.Values) Write(ev, sb, depth + 1);
                break;

            // -------- Sentencias --------
            case AssignStmt a:
                sb.AppendLine($"{ind}AssignStmt op={a.Op}");
                sb.AppendLine($"{ind}  target:");
                Write(a.Target, sb, depth + 2);
                sb.AppendLine($"{ind}  value:");
                Write(a.Value, sb, depth + 2);
                break;
            case IfStmt iff:
                sb.AppendLine($"{ind}IfStmt{(iff.IsSingleLine ? " [single-line]" : "")}");
                sb.AppendLine($"{ind}  then:");
                Write(iff.Then, sb, depth + 2);
                foreach (var ei in iff.ElseIfs)
                {
                    sb.AppendLine($"{ind}  elseif:");
                    Write(ei, sb, depth + 2);
                }
                if (iff.Else is not null)
                {
                    sb.AppendLine($"{ind}  else:");
                    foreach (var st in iff.Else) Write(st, sb, depth + 2);
                }
                break;
            case IfClause ic:
                sb.AppendLine($"{ind}IfClause");
                sb.AppendLine($"{ind}  condition:");
                Write(ic.Condition, sb, depth + 2);
                sb.AppendLine($"{ind}  body:");
                foreach (var st in ic.Body) Write(st, sb, depth + 2);
                break;
            case SwitchStmt sw:
                sb.AppendLine($"{ind}SwitchStmt");
                sb.AppendLine($"{ind}  subject:");
                Write(sw.Subject, sb, depth + 2);
                foreach (var cc in sw.Cases) Write(cc, sb, depth + 1);
                if (sw.Default is not null)
                {
                    sb.AppendLine($"{ind}  default:");
                    foreach (var st in sw.Default) Write(st, sb, depth + 2);
                }
                break;
            case CaseClause cc:
                sb.AppendLine($"{ind}CaseClause");
                sb.AppendLine($"{ind}  values:");
                foreach (var e in cc.Values) Write(e, sb, depth + 2);
                sb.AppendLine($"{ind}  body:");
                foreach (var st in cc.Body) Write(st, sb, depth + 2);
                break;
            case WhileStmt w:
                sb.AppendLine($"{ind}WhileStmt{(w.IsSingleLine ? " [single-line]" : "")}");
                sb.AppendLine($"{ind}  condition:");
                Write(w.Condition, sb, depth + 2);
                sb.AppendLine($"{ind}  body:");
                foreach (var st in w.Body) Write(st, sb, depth + 2);
                break;
            case DoLoopStmt dl:
                sb.AppendLine($"{ind}DoLoopStmt{(dl.Condition is null ? " [infinite]" : "")}");
                sb.AppendLine($"{ind}  body:");
                foreach (var st in dl.Body) Write(st, sb, depth + 2);
                if (dl.Condition is not null)
                {
                    sb.AppendLine($"{ind}  while:");
                    Write(dl.Condition, sb, depth + 2);
                }
                break;
            case ForStmt fs:
                sb.AppendLine($"{ind}ForStmt iter={fs.IteratorName}{(fs.IsSingleLine ? " [single-line]" : "")}");
                Write(fs.Range, sb, depth + 1);
                sb.AppendLine($"{ind}  body:");
                foreach (var st in fs.Body) Write(st, sb, depth + 2);
                break;
            case ForNumericRange fnr:
                sb.AppendLine($"{ind}NumericRange");
                sb.AppendLine($"{ind}  from:"); Write(fnr.From, sb, depth + 2);
                sb.AppendLine($"{ind}  to:");   Write(fnr.To,   sb, depth + 2);
                if (fnr.Step is not null) { sb.AppendLine($"{ind}  step:"); Write(fnr.Step, sb, depth + 2); }
                break;
            case ForInRange fir:
                sb.AppendLine($"{ind}InRange");
                Write(fir.Iterable, sb, depth + 1);
                break;
            case TryStmt tr:
                sb.AppendLine($"{ind}TryStmt");
                sb.AppendLine($"{ind}  body:");
                foreach (var st in tr.Body) Write(st, sb, depth + 2);
                foreach (var cl in tr.Catches) Write(cl, sb, depth + 1);
                if (tr.Finally is not null)
                {
                    sb.AppendLine($"{ind}  finally:");
                    foreach (var st in tr.Finally) Write(st, sb, depth + 2);
                }
                break;
            case CatchClause cl:
                string vn = cl.VarName ?? "_";
                string tn = cl.ExceptionType is null ? "*" : cl.ExceptionType;
                sb.AppendLine($"{ind}Catch (var={vn}, type={tn})");
                foreach (var st in cl.Body) Write(st, sb, depth + 1);
                break;
            case ReturnStmt r:
                sb.AppendLine($"{ind}Return{(r.Value is null ? " (void)" : "")}");
                if (r.Value is not null) Write(r.Value, sb, depth + 1);
                break;
            case ThrowStmt th:
                sb.AppendLine($"{ind}Throw");
                Write(th.Value, sb, depth + 1);
                break;
            case PrintStmt pr2:
                sb.AppendLine($"{ind}Print");
                foreach (var it in pr2.Items)
                {
                    sb.AppendLine($"{ind}  item sep={it.LeadingSep}");
                    if (it.Expr is not null) Write(it.Expr, sb, depth + 2);
                }
                break;
            case BreakStmt:    sb.AppendLine($"{ind}Break"); break;
            case ContinueStmt: sb.AppendLine($"{ind}Continue"); break;
            case ExprStmt es:
                sb.AppendLine($"{ind}ExprStmt");
                Write(es.Expr, sb, depth + 1);
                break;

            // -------- Expresiones --------
            case IntLitExpr i:    sb.AppendLine($"{ind}Int    {i.Value}"); break;
            case FloatLitExpr fl: sb.AppendLine($"{ind}Float  {fl.Value}"); break;
            case StringLitExpr s2:sb.AppendLine($"{ind}String \"{Escape(s2.Value)}\""); break;
            case BoolLitExpr b:   sb.AppendLine($"{ind}Bool   {b.Value}"); break;
            case NullLitExpr:     sb.AppendLine($"{ind}Null"); break;
            case ThisExpr:        sb.AppendLine($"{ind}This"); break;
            case SuperExpr:       sb.AppendLine($"{ind}Super"); break;
            case FieldExpr:       sb.AppendLine($"{ind}Field"); break;
            case SuperCallExpr sc:
                sb.AppendLine($"{ind}SuperCall");
                foreach (var a in sc.Args) Write(a, sb, depth + 1);
                break;
            case IdentifierExpr id:
                sb.AppendLine($"{ind}Ident  {id.Name}");
                break;
            case ArrayLitExpr al:
                sb.AppendLine($"{ind}ArrayLit ({al.Elements.Count})");
                foreach (var e in al.Elements) Write(e, sb, depth + 1);
                break;
            case ParenExpr pe:
                sb.AppendLine($"{ind}Paren");
                Write(pe.Inner, sb, depth + 1);
                break;
            case MemberAccessExpr ma:
                sb.AppendLine($"{ind}Member .{ma.Member}");
                Write(ma.Target, sb, depth + 1);
                break;
            case IndexExpr ix:
                sb.AppendLine($"{ind}Index");
                sb.AppendLine($"{ind}  target:");
                Write(ix.Target, sb, depth + 2);
                sb.AppendLine($"{ind}  index:");
                Write(ix.Index, sb, depth + 2);
                break;
            case CallExpr ce:
                sb.AppendLine($"{ind}Call");
                sb.AppendLine($"{ind}  callee:");
                Write(ce.Callee, sb, depth + 2);
                if (ce.Args.Count > 0)
                {
                    sb.AppendLine($"{ind}  args:");
                    foreach (var a in ce.Args) Write(a, sb, depth + 2);
                }
                break;
            case UnaryExpr u:
                sb.AppendLine($"{ind}Unary {u.Op}");
                Write(u.Operand, sb, depth + 1);
                break;
            case BinaryExpr bin:
                sb.AppendLine($"{ind}Binary {bin.Op}");
                Write(bin.Left, sb, depth + 1);
                Write(bin.Right, sb, depth + 1);
                break;

            default:
                sb.AppendLine($"{ind}<{node?.GetType().Name ?? "?"}>");
                break;
        }
    }

    private static string Mods(bool isPublic, bool isFinal)
    {
        var sb = new StringBuilder();
        if (isPublic) sb.Append(" [public]");
        if (isFinal)  sb.Append(" [final]");
        return sb.ToString();
    }

    private static string Escape(string s) =>
        s.Replace("\\", "\\\\")
         .Replace("\"", "\\\"")
         .Replace("\n", "\\n")
         .Replace("\t", "\\t")
         .Replace("\r", "\\r");
}

File: lexer-java/src/main/java/basicplus/frontend/AstPrinter.java
// ============================================================
// AstPrinter.java
// Vuelca un AST de BASICPLUS en formato indentado tipo árbol.
// Port a Java 8 del AstPrinter.cs.
// ============================================================
package basicplus.frontend;

import basicplus.frontend.Ast.*;

public final class AstPrinter {

    private AstPrinter() {}

    public static String print(Object node) {
        StringBuilder sb = new StringBuilder();
        write(node, sb, 0);
        return sb.toString();
    }

    private static String indent(int depth) {
        char[] buf = new char[depth * 2];
        for (int i = 0; i < buf.length; i++) buf[i] = ' ';
        return new String(buf);
    }

    private static String mods(boolean isPublic, boolean isFinal) {
        StringBuilder sb = new StringBuilder();
        if (isPublic) sb.append(" [public]");
        if (isFinal)  sb.append(" [final]");
        return sb.toString();
    }

    private static String escape(String s) {
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\t", "\\t")
                .replace("\r", "\\r");
    }

    private static void write(Object node, StringBuilder sb, int depth) {
        String ind = indent(depth);

        if (node == null) { sb.append(ind).append("<null>\n"); return; }

        // ---------- Módulo / imports ----------
        if (node instanceof ModuleNode) {
            ModuleNode m = (ModuleNode) node;
            sb.append(ind).append("Module '").append(m.name).append("'  @ ")
                    .append(m.line).append(":").append(m.column).append("\n");
            for (ImportNode imp : m.imports) write(imp, sb, depth + 1);
            for (ITopLevelDecl d : m.defs)   write(d,   sb, depth + 1);
            return;
        }
        if (node instanceof ImportNode) {
            ImportNode i = (ImportNode) node;
            sb.append(ind).append("Import ").append(String.join(".", i.path)).append("\n");
            return;
        }

        // ---------- Tipos ----------
        if (node instanceof SimpleTypeRef) {
            sb.append(ind).append("Type ").append(((SimpleTypeRef) node).name).append("\n");
            return;
        }
        if (node instanceof ArrayTypeRef) {
            ArrayTypeRef a = (ArrayTypeRef) node;
            sb.append(ind).append("ArrayType\n");
            write(a.element, sb, depth + 1);
            if (a.size != null) {
                sb.append(ind).append("  size:\n");
                write(a.size, sb, depth + 2);
            }
            return;
        }

        // ---------- Helpers ----------
        if (node instanceof DeclName) {
            DeclName dn = (DeclName) node;
            String prefix = dn.isStatic() ? dn.classQualifier + "." : "";
            sb.append(ind).append("DeclName ").append(prefix).append(dn.name).append("\n");
            return;
        }
        if (node instanceof Param) {
            Param p = (Param) node;
            sb.append(ind).append("Param ").append(p.name).append("\n");
            write(p.type, sb, depth + 1);
            return;
        }
        if (node instanceof EnumValue) {
            EnumValue ev = (EnumValue) node;
            String val = (ev.explicitValue != null) ? " := " + ev.explicitValue : "";
            sb.append(ind).append("EnumValue ").append(ev.name).append(val).append("\n");
            return;
        }
        if (node instanceof GetterDef) {
            GetterDef g = (GetterDef) node;
            sb.append(ind).append("Getter\n");
            for (IStmt s : g.body) write(s, sb, depth + 1);
            return;
        }
        if (node instanceof SetterDef) {
            SetterDef s = (SetterDef) node;
            sb.append(ind).append("Setter (param=").append(s.paramName).append(")\n");
            for (IStmt st : s.body) write(st, sb, depth + 1);
            return;
        }

        // ---------- Declaraciones ----------
        if (node instanceof ConstDecl) {
            ConstDecl c = (ConstDecl) node;
            sb.append(ind).append("ConstDecl").append(mods(c.isPublic, false)).append("\n");
            write(c.name, sb, depth + 1);
            if (c.type != null) write(c.type, sb, depth + 1);
            sb.append(ind).append("  value:\n");
            write(c.value, sb, depth + 2);
            return;
        }
        if (node instanceof VarDecl) {
            VarDecl v = (VarDecl) node;
            sb.append(ind).append("VarDecl").append(mods(v.isPublic, false)).append("\n");
            for (DeclName n : v.names) write(n, sb, depth + 1);
            write(v.type, sb, depth + 1);
            if (v.init != null) {
                sb.append(ind).append("  init:\n");
                write(v.init, sb, depth + 2);
            }
            return;
        }
        if (node instanceof FuncDef) {
            FuncDef f = (FuncDef) node;
            sb.append(ind).append("FuncDef").append(mods(f.isPublic, f.isFinal)).append("\n");
            write(f.name, sb, depth + 1);
            for (Param p : f.params) write(p, sb, depth + 1);
            if (f.returnType != null) {
                sb.append(ind).append("  returns:\n");
                write(f.returnType, sb, depth + 2);
            }
            sb.append(ind).append("  body:\n");
            for (IStmt st : f.body) write(st, sb, depth + 2);
            return;
        }
        if (node instanceof PropertyDef) {
            PropertyDef pr = (PropertyDef) node;
            sb.append(ind).append("PropertyDef")
                    .append(mods(pr.isPublic, pr.isFinal))
                    .append(pr.isShortForm ? " [short]" : "").append("\n");
            write(pr.name, sb, depth + 1);
            write(pr.type, sb, depth + 1);
            if (pr.init != null) {
                sb.append(ind).append("  init:\n");
                write(pr.init, sb, depth + 2);
            }
            if (pr.getter != null) write(pr.getter, sb, depth + 1);
            if (pr.setter != null) write(pr.setter, sb, depth + 1);
            return;
        }
        if (node instanceof ClassDef) {
            ClassDef cd = (ClassDef) node;
            sb.append(ind).append("ClassDef '").append(cd.name).append("'");
            if (cd.baseClass != null) sb.append(" extends ").append(cd.baseClass);
            sb.append(mods(cd.isPublic, false)).append("\n");
            for (ITopLevelDecl m : cd.members) write(m, sb, depth + 1);
            return;
        }
        if (node instanceof EnumDef) {
            EnumDef en = (EnumDef) node;
            sb.append(ind).append("EnumDef '").append(en.name).append("'")
                    .append(mods(en.isPublic, false)).append("\n");
            for (EnumValue ev : en.values) write(ev, sb, depth + 1);
            return;
        }

        // ---------- Sentencias ----------
        if (node instanceof AssignStmt) {
            AssignStmt a = (AssignStmt) node;
            sb.append(ind).append("AssignStmt op=").append(a.op).append("\n");
            sb.append(ind).append("  target:\n");
            write(a.target, sb, depth + 2);
            sb.append(ind).append("  value:\n");
            write(a.value, sb, depth + 2);
            return;
        }
        if (node instanceof IfStmt) {
            IfStmt iff = (IfStmt) node;
            sb.append(ind).append("IfStmt").append(iff.isSingleLine ? " [single-line]" : "").append("\n");
            sb.append(ind).append("  then:\n");
            write(iff.then_, sb, depth + 2);
            for (IfClause ei : iff.elseIfs) {
                sb.append(ind).append("  elseif:\n");
                write(ei, sb, depth + 2);
            }
            if (iff.else_ != null) {
                sb.append(ind).append("  else:\n");
                for (IStmt st : iff.else_) write(st, sb, depth + 2);
            }
            return;
        }
        if (node instanceof IfClause) {
            IfClause ic = (IfClause) node;
            sb.append(ind).append("IfClause\n");
            sb.append(ind).append("  condition:\n");
            write(ic.condition, sb, depth + 2);
            sb.append(ind).append("  body:\n");
            for (IStmt st : ic.body) write(st, sb, depth + 2);
            return;
        }
        if (node instanceof SwitchStmt) {
            SwitchStmt sw = (SwitchStmt) node;
            sb.append(ind).append("SwitchStmt\n");
            sb.append(ind).append("  subject:\n");
            write(sw.subject, sb, depth + 2);
            for (CaseClause cc : sw.cases) write(cc, sb, depth + 1);
            if (sw.defaultBody != null) {
                sb.append(ind).append("  default:\n");
                for (IStmt st : sw.defaultBody) write(st, sb, depth + 2);
            }
            return;
        }
        if (node instanceof CaseClause) {
            CaseClause cc = (CaseClause) node;
            sb.append(ind).append("CaseClause\n");
            sb.append(ind).append("  values:\n");
            for (IExpr e : cc.values) write(e, sb, depth + 2);
            sb.append(ind).append("  body:\n");
            for (IStmt st : cc.body) write(st, sb, depth + 2);
            return;
        }
        if (node instanceof WhileStmt) {
            WhileStmt w = (WhileStmt) node;
            sb.append(ind).append("WhileStmt").append(w.isSingleLine ? " [single-line]" : "").append("\n");
            sb.append(ind).append("  condition:\n");
            write(w.condition, sb, depth + 2);
            sb.append(ind).append("  body:\n");
            for (IStmt st : w.body) write(st, sb, depth + 2);
            return;
        }
        if (node instanceof DoLoopStmt) {
            DoLoopStmt dl = (DoLoopStmt) node;
            sb.append(ind).append("DoLoopStmt").append(dl.condition == null ? " [infinite]" : "").append("\n");
            sb.append(ind).append("  body:\n");
            for (IStmt st : dl.body) write(st, sb, depth + 2);
            if (dl.condition != null) {
                sb.append(ind).append("  while:\n");
                write(dl.condition, sb, depth + 2);
            }
            return;
        }
        if (node instanceof ForStmt) {
            ForStmt fs = (ForStmt) node;
            sb.append(ind).append("ForStmt iter=").append(fs.iteratorName)
                    .append(fs.isSingleLine ? " [single-line]" : "").append("\n");
            write(fs.range, sb, depth + 1);
            sb.append(ind).append("  body:\n");
            for (IStmt st : fs.body) write(st, sb, depth + 2);
            return;
        }
        if (node instanceof ForNumericRange) {
            ForNumericRange fnr = (ForNumericRange) node;
            sb.append(ind).append("NumericRange\n");
            sb.append(ind).append("  from:\n"); write(fnr.from, sb, depth + 2);
            sb.append(ind).append("  to:\n");   write(fnr.to,   sb, depth + 2);
            if (fnr.step != null) { sb.append(ind).append("  step:\n"); write(fnr.step, sb, depth + 2); }
            return;
        }
        if (node instanceof ForInRange) {
            ForInRange fir = (ForInRange) node;
            sb.append(ind).append("InRange\n");
            write(fir.iterable, sb, depth + 1);
            return;
        }
        if (node instanceof TryStmt) {
            TryStmt tr = (TryStmt) node;
            sb.append(ind).append("TryStmt\n");
            sb.append(ind).append("  body:\n");
            for (IStmt st : tr.body) write(st, sb, depth + 2);
            for (CatchClause cl : tr.catches) write(cl, sb, depth + 1);
            if (tr.finallyBody != null) {
                sb.append(ind).append("  finally:\n");
                for (IStmt st : tr.finallyBody) write(st, sb, depth + 2);
            }
            return;
        }
        if (node instanceof CatchClause) {
            CatchClause cl = (CatchClause) node;
            String vn = cl.varName == null ? "_" : cl.varName;
            String tn = cl.exceptionType == null ? "*" : cl.exceptionType;
            sb.append(ind).append("Catch (var=").append(vn).append(", type=").append(tn).append(")\n");
            for (IStmt st : cl.body) write(st, sb, depth + 1);
            return;
        }
        if (node instanceof ReturnStmt) {
            ReturnStmt r = (ReturnStmt) node;
            sb.append(ind).append("Return").append(r.value == null ? " (void)" : "").append("\n");
            if (r.value != null) write(r.value, sb, depth + 1);
            return;
        }
        if (node instanceof ThrowStmt) {
            ThrowStmt th = (ThrowStmt) node;
            sb.append(ind).append("Throw\n");
            write(th.value, sb, depth + 1);
            return;
        }
        if (node instanceof PrintStmt) {
            PrintStmt p = (PrintStmt) node;
            sb.append(ind).append("Print\n");
            for (PrintItem it : p.items) {
                sb.append(ind).append("  item sep=").append(it.leadingSep).append("\n");
                if (it.expr != null) write(it.expr, sb, depth + 2);
            }
            return;
        }
        if (node instanceof BreakStmt)    { sb.append(ind).append("Break\n");    return; }
        if (node instanceof ContinueStmt) { sb.append(ind).append("Continue\n"); return; }
        if (node instanceof ExprStmt) {
            ExprStmt es = (ExprStmt) node;
            sb.append(ind).append("ExprStmt\n");
            write(es.expr, sb, depth + 1);
            return;
        }

        // ---------- Expresiones ----------
        if (node instanceof IntLitExpr) {
            sb.append(ind).append("Int    ").append(((IntLitExpr) node).value).append("\n"); return;
        }
        if (node instanceof FloatLitExpr) {
            sb.append(ind).append("Float  ").append(((FloatLitExpr) node).value).append("\n"); return;
        }
        if (node instanceof StringLitExpr) {
            sb.append(ind).append("String \"").append(escape(((StringLitExpr) node).value)).append("\"\n"); return;
        }
        if (node instanceof BoolLitExpr) {
            sb.append(ind).append("Bool   ").append(((BoolLitExpr) node).value).append("\n"); return;
        }
        if (node instanceof NullLitExpr) { sb.append(ind).append("Null\n"); return; }
        if (node instanceof ThisExpr)    { sb.append(ind).append("This\n"); return; }
        if (node instanceof SuperExpr)   { sb.append(ind).append("Super\n"); return; }
        if (node instanceof FieldExpr)   { sb.append(ind).append("Field\n"); return; }
        if (node instanceof SuperCallExpr) {
            SuperCallExpr sc = (SuperCallExpr) node;
            sb.append(ind).append("SuperCall\n");
            for (IExpr a : sc.args) write(a, sb, depth + 1);
            return;
        }
        if (node instanceof IdentifierExpr) {
            sb.append(ind).append("Ident  ").append(((IdentifierExpr) node).name).append("\n"); return;
        }
        if (node instanceof ArrayLitExpr) {
            ArrayLitExpr al = (ArrayLitExpr) node;
            sb.append(ind).append("ArrayLit (").append(al.elements.size()).append(")\n");
            for (IExpr e : al.elements) write(e, sb, depth + 1);
            return;
        }
        if (node instanceof ParenExpr) {
            sb.append(ind).append("Paren\n");
            write(((ParenExpr) node).inner, sb, depth + 1);
            return;
        }
        if (node instanceof MemberAccessExpr) {
            MemberAccessExpr ma = (MemberAccessExpr) node;
            sb.append(ind).append("Member .").append(ma.member).append("\n");
            write(ma.target, sb, depth + 1);
            return;
        }
        if (node instanceof IndexExpr) {
            IndexExpr ix = (IndexExpr) node;
            sb.append(ind).append("Index\n");
            sb.append(ind).append("  target:\n");
            write(ix.target, sb, depth + 2);
            sb.append(ind).append("  index:\n");
            write(ix.index, sb, depth + 2);
            return;
        }
        if (node instanceof CallExpr) {
            CallExpr ce = (CallExpr) node;
            sb.append(ind).append("Call\n");
            sb.append(ind).append("  callee:\n");
            write(ce.callee, sb, depth + 2);
            if (!ce.args.isEmpty()) {
                sb.append(ind).append("  args:\n");
                for (IExpr a : ce.args) write(a, sb, depth + 2);
            }
            return;
        }
        if (node instanceof UnaryExpr) {
            UnaryExpr u = (UnaryExpr) node;
            sb.append(ind).append("Unary ").append(u.op).append("\n");
            write(u.operand, sb, depth + 1);
            return;
        }
        if (node instanceof BinaryExpr) {
            BinaryExpr b = (BinaryExpr) node;
            sb.append(ind).append("Binary ").append(b.op).append("\n");
            write(b.left, sb, depth + 1);
            write(b.right, sb, depth + 1);
            return;
        }
        if (node instanceof InstanceOfExpr) {
            InstanceOfExpr io = (InstanceOfExpr) node;
            sb.append(ind).append("InstanceOf ").append(io.typeName).append("\n");
            write(io.target, sb, depth + 1);
            return;
        }

        sb.append(ind).append("<").append(node.getClass().getSimpleName()).append(">\n");
    }
}
